tests/unittests/Foundation/NSThreadTests.m

```diff
@@ -77,6 +78,24 @@ - (void)countingloop {
     testController.stop = YES;
 }
 
+TEST(NSThread, ExternalCurrentThreadLeak) {
+    NSThread* current = nil;
+
+    std::thread t([&current]() {
+        // Grab a strong reference to NSThread.currentThread
+        // on a thread not created via the NSThread APIs
+        current = [[NSThread currentThread] retain];
+    });
+
+    t.join();
+
+    // Ensure the reference grabbed above is
+    // the only one remaining after thread exit
+    EXPECT_EQ(1, [current retainCount]);
```

Comment:
In the body of the lambda we're not creating the NSThread via alloc init. We just get a non-owning reference by the call to |currentThread|. The call to |currentThread| does happen to create an owning reference (because it's called from a thread not created via the NSThread API), but this is relinquished on thread exit via the thread_local StrongId destructor. 

I'll add some comments to the test case when I add in @DHowett-MSFT's suggested changes - likely sometime tonight. It might also be clearer to convert that test to ARC - but I had some difficulty attempting this initially.